### PR TITLE
[FLINK-8439] Add Flink shading to AWS credential provider s3 hadoop c…

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

This PR refactors S3 Hadoop and Presto file systems and adds Flink shading to AWS credential provider config.

## Brief change log

  - extract `AbstractS3FileSystemFactory` base class from `s3hadoop.` and `s3presto.S3FileSystemFactory`s
  - extract hadoop configuration logic into `HadoopConfigLoader` with Flink shading of certain Hadoop configs
  -  add Flink shading to AWS credential provider config of `S3FileSystemFactory`s

## Verifying this change

run unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by azagrebin** [comment]
cc @GJL @StephanEwen

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/AbstractS3FileSystemFactory.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.flink.core.fs.FileSystemFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+
+/** Base class for S3 file system factories. */
+public abstract class AbstractS3FileSystemFactory implements FileSystemFactory {
```
Comment (aljoscha) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/AbstractS3FileSystemFactory.java]: I don't like putting S3 specifics into the generic Hadoop FS package. We could call this one `AbstractHadoopFileSystemFactory`, leave out the `getScheme()` implementation, and drop mentions of S3  to make it properly independent of S3.

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/** This class lazily loads hadoop configuration from resettable Flink's configuration. */
+public class HadoopConfigLoader {
+	private static final Logger LOG = LoggerFactory.getLogger(HadoopConfigLoader.class);
+
+	private static final Set<String> PACKAGE_PREFIXES_TO_SHADE =
```
Comment (aljoscha) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java]: Same as my other comment, I would like this to not be S3 specific. By making this another constructor parameter we could do that.

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		}
+		return hadoopConfig;
+	}
+
+	private String fixHadoopConfig(String key, String value) {
+		return key != null && configKeysToShade.contains(key) ?
+			shadeAwsClassConfig(value) : value;
+	}
+
+	private String shadeAwsClassConfig(String awsCredProvider) {
```
Comment (aljoscha) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java]: This is not tied to AWS or the credentials provider, I think.

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+					LOG.debug("Adding Flink config entry for {} as {}={} to Hadoop config", key, newKey, value);
+				}
+			}
+		}
+		return hadoopConfig;
+	}
+
+	// mirror certain keys to make use more uniform across implementations
+	// with different keys
+	private org.apache.hadoop.conf.Configuration mirrorCertianHadoopConfig(
```
Comment (aljoscha) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java]: typo `Certian` -> `Certain`

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java (hunk 3 above)
Comment (azagrebin) [reply flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java]: true, overlooked

**Comment by azagrebin** [comment]
Thanks for the review @aljoscha, I removed s3 specifics from the base class

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/AbstractS3FileSystemFactory.java (hunk 1 above)
Comment (azagrebin) [reply flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/AbstractS3FileSystemFactory.java]: I also thought about it, just the s3 classes still have something in common and no other common package, but probably too little in common

**Comment by aljoscha** [comment]
Thanks! This looks good now. I'll merge once Travis is green. 👌

**Pull request closed without merging**